src/metaopt/io/database/mongodb.py

```diff
@@ -79,43 +79,94 @@ def is_locked(self):
+    def insert(self, collection_name, document_dictionaries):
+        """ Inserts a document/multiple documents
+
+        Keyword arguments:
+        collection_name -- the collection name
+        document_dictionaries -- the dictionary/dictionaries that would be inserted
+                                 inside the collection
+        Returns:
+        result -- an instance of InsertManyResult which has inserted_ids and acknowledged
+        """
 
         dbcollection = self.db[collection_name]
-        result = dbcollection.insert_many(query, dictionaries)
+        result = dbcollection.insert_many(document_dictionaries)
 
         return result
 
-    def update(self, collection_name, query, update_dictionary):
-        """
-        note here that update needs to have
-        MongoDB comparison operations
-        for example update=
-        {
-           "$set": {'hyperparam_name': 'hidden units',
-                    'hyperparam_type': 'int',
-                    'hyperparam_value': '200',
-                    }
-        }
+    def update(self, collection_name, query, update_dictionary,
+               selection, upsert=False, sort=None,
+               return_document=ReturnDocument.BEFORE):
+        """ Finds a document and modify it
+
+        Keyword arguments:
+        collection_name -- the collection name
+        query -- the filter for the
+        update_dictionary -- the dictionary/dictionaries that would be updated
+        selection -- the projection
+        upsert -- if True if the query doesn't match, it will insert a new document
+        sort -- set to default as None, you could do Ascending or Descending
+        return_document -- the result returned would be the document before or after
+                            the update
+        Returns:
+        result -- the document specified in return_document
         """
 
+        # see http://api.mongodb.com/python/current/api/pymongo/collection.html
         dbcollection = self.db[collection_name]
 
-        # see http://api.mongodb.com/python/current/api/pymongo/collection.html
-        # for more details
+        """
+        for more details see
+        http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.find_one_and_update
+        here we previously used update_one. Could define if we
+        find the value and update it, or if no match query/filter is found
+        we could insert the document by setting upsert=True
+        although a find_and_modify is exactly that we would want
+        but the interface is deprecated so we use find_and_update() instread
 
-        result = dbcollection.update_one(query, update_dictionary)
+        """
+
+        result = dbcollection.find_one_and_update(filter=query,
+                                         update=update_dictionary,
+                                         projection=selection,
+                                         upsert=upsert,
```

Comment:
`upsert` is a nice additon, but why limit ourselves to one update? Please don't make a new commit with update_many. :slightly_smiling_face:  I'd prefer we discuss it and we'll see if we settle on find_one_and_update or update_many.

@tsirif Any thoughts on this?

Reply:
We recently encountered a problem by doing a nested query inside a dictionary, and by the way `update` is deprecated http://api.mongodb.com/python/current/api/pymongo/collection.html (at the bottom of the page)

Reply:
update is deprecated but update_{one, many} are not.

Do you have an example for the problem? I'd like to understand it better. It's not clear to me that update_one rather than update_many would be the best fix for that kind of problem.

Reply:
Also you can not update two documents at once with a MongoDB query. You will always have to do that in two queries. You can of course set a value of a field to the same value, or increment with the same number, but you can not do two distinct updates in MongoDB with the same query. As a reference https://stackoverflow.com/questions/17207183/way-to-update-multiple-documets-with-different-values

Reply:
Yes sure. Different updates need different calls to update(). I don't see that as an argument against update_many. The only argument I see in favor of find_one_and_update is the atomicity of the operation. I'm not sure however it is worth the cost for flexibility.